modules/actor/sheet/actor-sheet.js

```diff
@@ -513,11 +513,31 @@ export default class ActorSheetWfrp4e extends ActorSheet {
             channel: {
               label: game.i18n.localize("Channel"),
               callback: btn => {
+                options.channelUntilSuccess = false;
                 this.actor.setupChannell(spell, options).then(setupData => {
                   this.actor.channelTest(setupData)
                 });
               }
             },
+            channelUntilSuccess: {
+              label: game.i18n.localize("DIALOG.ChannelUntilReady"),
+              callback: btn => {
+                options.channelUntilSuccess = true;
+                this.actor.setupChannell(spell, options).then(async setupData => {
+                  do {
+                    let testDuplicate = duplicate(setupData);
+                    let testObject = new setupData.rollClass(testDuplicate);
+                    await this.actor.channelTest(testObject)
+                    if (testObject.result.minormis || testObject.result.majormis || testObject.result.catastrophicmis) {
+                      break;
+                    }
+                    if (testObject.item.cn.SL == testObject.item.cn.value) {
+                      break;
+                    }
+                  } while (true)
+                });
+              }
+            }
```

Comment:
I think I would prefer this as a checkbox in the channel dialog, what do you think?

Reply:
initially i did exatly that, but this would require to make an extra change in channel callback, where i dont have access to testObject and also all those setup<Test>().then(setupData => {}) are not awaited. Therefore it would require to break nice pattern that you have and extra button removes this problem. Also, i think it simply looks better (small chance that someone will do that by mistake).

Reply:
Maybe I'm misunderstanding, but this seems fine? I suppose we disagree because I don't really like the look of the button (especially because it expands the height of them all).

```js
            channel: {
              label: game.i18n.localize("Channel"),
              callback: async btn => {
                let test = await this.actor.setupChannell(spell, options)
                if (test.context.channelUntilSuccess) // Added via a dialog checkbox
                {
                   do {
                    let testData = duplicate(test);
                    let testDuplicate = new test.rollClass(testObject)
                    if (test.result.minormis || test.result.majormis || test.result.catastrophicmis) {
                      break;
                    }
                    if (test.item.cn.SL == test.item.cn.value) {
                      break;
                    }
                  } while (true)
                }
                else 
                   test.roll()
                }
              }
```

Regarding accidental activating this, perhaps we can remove this option if a combat is active?

Also, shouldn't the CN == SL check be the condition for the loop? Such as
`while (test.item.cn.SL < test.item.cn.value)`

Reply:
yeah, but this is excatly the change - any other callback isnt async and setupChannel or any other setupSkill/Weapon etc is never awaited, so this is the pattern i didnt want to break. Maybe they should, but they arent. i.e. those (note, that this example is not accidental, as some of the actions are async and some arent.)

```javascript
  _onWeaponNameClick(ev) {
    ev.preventDefault();
    let itemId = $(ev.currentTarget).parents(".item").attr("data-item-id");
    let weapon = this.actor.items.get(itemId)
    if (weapon) this.actor.setupWeapon(weapon).then(setupData => {
      if (!setupData.abort)
        this.actor.weaponTest(setupData)
    })
  }
  async _onUnarmedClick(ev) {
    ev.preventDefault();
    let unarmed = game.wfrp4e.config.systemItems.unarmed
    this.actor.setupWeapon(unarmed).then(setupData => {
      this.actor.weaponTest(setupData)
    })
  }
``` 

As for do{ } while (true) - i thought that it will be more readable, as we have two break conditions and with miscasts included, the while (condition) would be very long.

Reply:
So this is actually an old pattern from before I gave Tests their own classes. Notice how `basicTest` and `weaponTest` just simply call `test.roll()`

`setupData` shouldn't really be called that anymore because it's a test object when it used to be just its data. 

I didn't change it because I didn't want to break anything at the time. And those functions exist simply for backwards compatibility (mainly with effects)

Regardless the new pattern:
```js
this.actor.setupWeapon(weapon).then(test => test.roll())
```
I think is cleaner than 
```js
this.actor.setupWeapon(unarmed).then(setupData => this.actor.weaponTest(setupData))
```

Reply:
i will experiment on that in this case, i need to dig ia bit more into it, to understand the consequences when the test is awaited and when not.

Reply:
The test? or the setup? 

The test is awaited because rolling is async

The setup is awaited because it returns a Promise that is only resolved when the dialog is submitted, see: https://github.com/moo-man/WFRP4e-FoundryVTT/blob/master/modules/actor/actor-wfrp4e.js#L553-L571

Not sure if that clears up anything but let me know if you have any questions